algorithms/linfa-nn/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+use linfa::Float;
+use ndarray::{Array2, ArrayView1};
+
+pub mod kdtree;
+pub mod linear;
+
+pub(crate) type Point<'a, F> = ArrayView1<'a, F>;
+
+pub trait NearestNeighbour<F: Float> {
+    // Returns nearest in order. Might want wrap in result or return iterator
+    fn k_nearest<'b>(&self, point: Point<'b, F>, k: usize) -> Vec<Point<F>>;
+
+    // Does not have any particular order, though some algorithms may returns these in order of
+    // distance.
+    fn within_range<'b>(&self, point: Point<'b, F>, range: F) -> Vec<Point<F>>;
+}
+
+pub trait NearestNeighbourBuilder<F: Float> {
+    fn from_batch<'a>(&self, batch: &'a Array2<F>) -> Box<dyn 'a + NearestNeighbour<F>>;
```

Comment:
can you use an associated type here? like:
```rust
pub trait NearestNeighbourBuilder<'a, F: Float> {
    type Object: NearestNeighbour<F>;

    fn from_batch(&self, batch: &'a Array2<F>) -> Self::Object;
}

impl<'a, F: Float> NearestNeighbourBuilder<'a, F> for KdTreeBuilder<F> {
    type Object = KdTree<'a, F>;

    fn from_batch(&self, batch: &'a Array2<F>) -> Self::Object {
        KdTree::from_batch(batch)
    }
}
 ```

Reply:
Using an associated type would force the builder trait to have a lifetime, which is something I want to avoid because it complicates trait bounds and type declarations. I also want `from_batch` to return a trait object so APIs using nearest distance can dispatch to the desired algorithm at runtime.

Reply:
for me static dispatch is more important than the extra lifetime in the trait definition. But the function is called so few that it does not make a real difference

Reply:
If the builder trait has a lifetime then how would you even use it as a trait bound? I don't think you'd be able to use it inside a struct because its lifetime would have to last as long as the struct, so the lifetime of any `ArrayView1` passed into the algorithm would also need to last as long as the struct.

Reply:
the trait itself has not lifetime attached to the builder object, it only connects the lifetime of the returned `Self::Object` to that of the input batch. You would run into the problems you've described when `KdTreeBuilder<'a, F>` is parametrized

Reply:
I prototyped your trait design and got it working, but I had to add `where for<'a> B: NearestNeighbourBuilder<'a, f64>` to the signatures of the shared test functions. I suspect that similar clauses are needed to pass builders around generically in real-world use cases.